### PR TITLE
Refactor WelcomeNotification::Generator to ::Dispatcher

### PR DESCRIPTION
**Pull request by maestromac**
## What type of PR is this? (check all applicable)
- [x] Refactor

## Description
The goal of this PR is the following
- Extract the welcome notification logic into its own class called `Introduction`
- Change Generator to a module and renamed it, Dispatcher. I think it just makes more sense since it will decide what form of notification to call.
- This makes it really easy to add additional types of notifications.

The class names (`Introduction`, `Authentication`, etc..) is pending. Please feel free to suggest a better name.

side-note: I think we came come up with a more short and semantic name for `WelcomeNotifiations`
## Related Tickets & Documents
n/a
## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
n/a
## Added tests?
- [x] yes
## Added to documentation?
- [x] no documentation needed
## [optional] Are there any post deployment tasks we need to perform?
n/a

### CODE REVIEW

File: app/services/broadcasts/welcome_notification/introduction.rb [hunk 1]
```diff
@@ -1,37 +1,36 @@
...
-      end
+      private
 
       def commented_on_welcome_thread?
         welcome_thread = Article.admin_published_with("welcome").first
         Comment.where(commentable: welcome_thread, user: user).any?
       end
 
-      private
+      def received_notification?
```
Comment (vaidehijoshi) [inline app/services/broadcasts/welcome_notification/introduction.rb]: good call on making this `private`! i must have missed that it wasn't when i last reviewed this file

File: spec/services/broadcasts/welcome_notification/introduction_spec.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    let!(:welcome_broadcast) { create(:welcome_broadcast, :active) }
+
+    before do
+      allow(User).to receive(:mascot_account).and_return(mascot_account)
+      SiteConfig.staff_user_id = mascot_account.id
+    end
+
+    after do
+      SiteConfig.staff_user_id = 1
+    end
```
Comment (vaidehijoshi) [inline spec/services/broadcasts/welcome_notification/introduction_spec.rb]: something to note is that we'll need to do this setup across all welcome notifications

File: app/services/broadcasts/welcome_notification/introduction.rb [hunk 3]
```diff
@@ -1,37 +1,36 @@
...
+    class Introduction
+      attr_reader :user
+
+      def initialize(user)
+        @user = user
       end
 
-      def self.call(*args)
-        new(*args).call
+      def self.send(*args)
```
Comment (vaidehijoshi) [inline app/services/broadcasts/welcome_notification/introduction.rb]: 👍 good change!

**Comment by vaidehijoshi** [comment]
> side-note: I think we came come up with a more short and semantic name for WelcomeNotifiations

Could do something like `Broadcasts::Welcome::Dispatcher` instead

File: app/services/broadcasts/welcome_notification/introduction.rb (hunk 3 above)
Comment (juliannatetreault) [reply app/services/broadcasts/welcome_notification/introduction.rb]: Agreed!

**Comment by juliannatetreault** [comment]
> > side-note: I think we came come up with a more short and semantic name for WelcomeNotifiations
> 
> Could do something like `Broadcasts::Welcome::Dispatcher` instead

I'm pro this change. 👍

**Comment by maestromac** [comment]
Thanks for the feedback @vaidehijoshi . Fair point, I think I went down this path because I can envision the generator file to be cluttered very fast. I should definitely not get too ahead of myself. I'll close this off and refer to it when and if the time comes

**Comment by vaidehijoshi** [comment]
@maestromac No problem; thanks for being so open to hearing it! 😄 

I definitely agree that if the generator starts to be too cluttered or if the logic for notifications starts to diverge quite a bit, we should re-think this. I think we could definitely come back to this approach if we ever hit that scenario down the road!!

**Pull request closed without merging**